### PR TITLE
Fix parsing of block versions.

### LINKED ISSUE #307: Investigate version field in block 434873
The [Zcash specification]( https://zips.z.cash/protocol/protocol.pdf#blockheader) says that the current and only `version` field for block headers is `4`.  When trying to download and parse the chain, we fail at block 434873, due to a bad version field. Investigating with `zcash-cli` reveals a strange version field, not just a problem with our code:

```
~/c/z/zcashd (master|✔) $ ./src/zcash-cli getblock 434872 | jq '.version'
4
~/c/z/zcashd (master|✔) $ ./src/zcash-cli getblock 434873 | jq '.version'
536870912
~/c/z/zcashd (master|✔) $ ./src/zcash-cli getblock 434874 | jq '.version'
4
```

### LINKED ISSUE #308: Change block header parsing to relax version requirements.
#307 fails because one block has a bad version field. The spec says that the only defined version for Zcash blocks is 4; apparently this does not mean that the only *allowed* version is 4.  Per https://github.com/ZcashFoundation/zebra/issues/307#issuecomment-597287298 we should check that the high bit is unset and allow any random number to be included in the `version` field.  This seems like it makes the `version` field useless for actually controlling the block version.

### PR DESCRIPTION
**Pull request by hdevalence**
Closes #307 #308

### CODE REVIEW

File: zebra-chain/src/block.rs [hunk 1]
```diff
@@ -67,6 +67,18 @@ impl ZcashDeserialize for BlockHeaderHash {
     }
 }
 
+impl std::str::FromStr for BlockHeaderHash {
+    type Err = SerializationError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut bytes = [0; 32];
+        if hex::decode_to_slice(s, &mut bytes[..]).is_err() {
```
Comment (dconnolly) [inline zebra-chain/src/block.rs:74]: smart 👍

File: zebra-chain/src/block.rs [hunk 2]
```diff
@@ -119,8 +141,7 @@ pub struct BlockHeader {
 
 impl ZcashSerialize for BlockHeader {
     fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
-        // "The current and only defined block version number for Zcash is 4."
-        writer.write_u32::<LittleEndian>(4)?;
```
Comment (dconnolly) [inline zebra-chain/src/block.rs:123]: :single-tear:

File: zebra-chain/src/block/test_vectors.rs [hunk 3]
```diff
@@ -105,4 +105,5 @@ lazy_static! {
...
...000400008085202f8903c88f79329b0c4eae2777582d304c0e976d0febebc116bf8b35d5c8ce5cb5cd324c0000006b483045022100d17d5dc13d1eff9358e53cf3b74bc6550b2fcd1b4d5dc50fc143f19d11098a5d02206990e0b009814b5d5b14e2923406cdaf6950e766302eb61b7047e7671fd2a21a01210317ad10576b806c1824c46e5517e979e7a161b76315ba93e6fca5a3b4bf66f5e4feffffff6c2877d3879245810ebb32a2664a708677e20b30e1e5be5516c5e024af968f96020000006b483045022100e91f329f1dfa5e777c3db5204b05916c48e049213a11084356b66033a7557c7a02207adc7ddfc736be69840274f1ec94e3657614a2683a0a832a06b11bcff4893196012103d8c7454c1b13457d66980610122fc84e2ca876c251f65b4f94211e75955080d9feffffff59152d5ad96891a193920e241bad45efdbe153d7c5b0d4652526af5405b3de37000000006a47304402204fa72cabb5ceff3b3da9ebe1127c441b9840acd4b8c7e5a3599d8f28687d659a02206506df0f7ac1e0d331cb66edec89c92fce4d129234784bf0c007a5a5d526b879012103a633a7efb4b409a7e2135ae87a2a800450b9d91f1f28d69072b273f318e611bffeffffff03de3812000000000017a91439dec991976767ded3f51515033201049401710d8733231a00000000001976a914c743b8321f3116a169069aaf9f05d091ec6afe4b88ac809a10000000000017a91430b3cce69378a7c661368c669e135f951f3e63d987aea20600cda206000000000000000000000000") .expect("Block bytes are in valid hex representation");
```
Comment (dconnolly) [inline zebra-chain/src/block/test_vectors.rs:108]: 👍

File: zebra-chain/src/block/tests.rs [hunk 4]
```diff
@@ -16,6 +16,7 @@ impl Arbitrary for BlockHeader {
 
     fn arbitrary_with(_args: ()) -> Self::Strategy {
         (
+            (4u32..2_147_483_647u32),
```
Comment (dconnolly) [inline zebra-chain/src/block/tests.rs:19]: 👍

File: zebra-chain/src/block/tests.rs [hunk 5]
```diff
@@ -106,16 +110,17 @@ fn deserialize_blockheader() {
...
+    Block::zcash_deserialize(&test_vectors::BLOCK_MAINNET_1_BYTES[..])
         .expect("block test vector should deserialize");
-    println!("{:#?}", block);
     // https://explorer.zcha.in/blocks/415000
-    let block = Block::zcash_deserialize(&test_vectors::BLOCK_MAINNET_415000_BYTES[..])
+    Block::zcash_deserialize(&test_vectors::BLOCK_MAINNET_415000_BYTES[..])
+        .expect("block test vector should deserialize");
+    // https://explorer.zcha.in/blocks/434873
+    // this one has a bad version field
+    Block::zcash_deserialize(&test_vectors::BLOCK_MAINNET_434873_BYTES[..])
```
Comment (dconnolly) [inline zebra-chain/src/block/tests.rs:122]: 👍

**Review by dconnolly — APPROVED** [review]
🚣‍♂️

**Review by daira — COMMENTED** [review]
ACK

**Pull request merged**